psiturk/api/__init__.py

```diff
@@ -277,20 +193,335 @@ def post(self):
+    # POST: Returns the full list of workers from the local database
+    #  codeversion: the codeversion for which to retrieve workers for
+    def post(self):
+        worker_ids = request.json['worker_ids']
+        codeversion = request.json['codeversion']
+        query = Participant.query
+        if len(worker_ids) > 0:
+            query = query.filter(Participant.workerid.in_(worker_ids))
+        if codeversion:
+            query = query.filter(Participant.codeversion == codeversion)
+        _return = query.all()
+        return [p.toAPIData() for p in _return]
+
+class CampaignsList(Resource):
+
+    # POST: Returns a list of campaigns from the local database
+    #  codeversion: the codeversion for which to retrieve campaigns for
+    def post(self):
+        codeversion = request.json['codeversion']
+        query = Campaign.query.filter(Campaign.mode == services_manager.mode)
+        if codeversion:
+            query = query.filter(Campaign.codeversion == services_manager.codeversion)
+        _return = query.all()
+        return _return
 
-api.add_resource(Hits, '/hit/<hit_id>')
-api.add_resource(HitList, '/hits/', '/hits/<status>')
-api.add_resource(HitsAction, '/hits/action/<action>')
+class CampaignsAction(Resource):
+
+    # POST: Actions for campaigns
+    def post(self, action=None):
+
+        # ACTION: Create a campaign
+        #  goal: the target number of participants to hit
+        #  minutes_between_rounds: how many minutes to wait between rounds
+        #  assignments_between_rounds: how many assignments per round (9)
+        #  hit_reward: the reward for the HIT
+        #  hit_duration_hours: how long the HIT lasts before a user is unable to submit
+        if action == 'create':
+            goal = request.json['goal']
+            minutes_between_rounds = request.json['minutes_between_rounds']
+            assignments_per_round = request.json['assignments_per_round']
+            hit_reward = request.json['hit_reward']
+            hit_duration_hours = request.json['hit_duration_hours']
+            # check if one already exists that is active...
+            if Campaign.active_campaign_exists():
+                raise APIException(
+                    message='Active campaign already exists. Cancel that campaign first in order to create a new one.')
+            codeversion = services_manager.codeversion
+            mode = services_manager.mode
+            _return = Campaign.launch_new_campaign(
+                codeversion=codeversion, 
+                mode=mode, 
+                goal=goal, 
+                minutes_between_rounds=minutes_between_rounds,
+                assignments_per_round=assignments_per_round, 
+                hit_reward=hit_reward,
+                hit_duration_hours=hit_duration_hours)
+            return _return, 201
+
+
+
+class ServicesManager(Resource):
+
+    # POST: Set the services manager mode
+    #  mode: live or sandbox, the desired mode
+    def post(self):
```

Comment:
I'm uncomfortable with this -- I think the api shouldn't have any inherent concept of a session. The dashboard, on the other hand, can conceptually know about a session.

Reply:
On the other hand, my argument above probably means that the api shouldn't know about any services_manager either.